sparse.mm backward: performance improvements

nikitaved opened the pull request:
`torch.sparse.mm` - faster and without syncs in "most" cases.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #99298
* #99043
* #98490
* #95165
* __->__ #94991



cc @alexsamardzic @pearu @cpuhrsch @amjames @bhosmer @ezyang @albanD @zou3519 @gqchen @soulitzer @Lezcano @Varal7

nikitaved committed:
sparse.mm backward: performance improvements

[ghstack-poisoned]

nikitaved committed:
Update on "sparse.mm backward: performance improvements"




[ghstack-poisoned]

nikitaved commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -1370,19 +1370,13 @@ Tensor sparse_sparse_matmul_backward(
   if (grad_order == 0) {
     auto a_grad = _sparse_sparse_matmul(grad, b.conj().t());
-    return a_grad.mul(mask_ones_like(a.coalesce()));
+    return a_grad.sparse_mask(a.coalesce());
```
It is still WIP, we can remove `coalesce` altogether.

nikitaved replied:
Done, will push soon.

nikitaved committed:
Update on "sparse.mm backward: performance improvements"




[ghstack-poisoned]

nikitaved committed:
Update on "sparse.mm backward: performance improvements"




[ghstack-poisoned]

nikitaved commented:
@albanD , @soulitzer , could you please also have a look?

nikitaved committed:
Update on "sparse.mm backward: performance improvements"




[ghstack-poisoned]

nikitaved committed:
Update on "sparse.mm backward: performance improvements"

`torch.sparse.mm` - faster and without syncs in "most" cases.




[ghstack-poisoned]

nikitaved committed:
Update on "sparse.mm backward: performance improvements"

`torch.sparse.mm` - faster and without syncs in "most" cases.




[ghstack-poisoned]

nikitaved committed:
Update on "sparse.mm backward: performance improvements"

`torch.sparse.mm` - faster and without syncs in "most" cases.




[ghstack-poisoned]

nikitaved committed:
Update on "sparse.mm backward: performance improvements"

`torch.sparse.mm` - faster and without syncs in "most" cases.




[ghstack-poisoned]

nikitaved committed:
Update on "sparse.mm backward: performance improvements"

`torch.sparse.mm` - faster and without syncs in "most" cases.




cc alexsamardzic pearu cpuhrsch amjames bhosmer ezyang albanD zou3519 gqchen soulitzer Lezcano Varal7

[ghstack-poisoned]

Skylion007 commented on aten/src/ATen/native/sparse/SparseTensor.cpp:
```diff
@@ -744,6 +746,65 @@ SparseTensor _coalesce_sparse_cpu(const SparseTensor& self) {
+      // NOTE: res is not necessariy coalesced, but it is sorted.
+      // We mark it as "coalesced" to skip sorting in the intersection kernel.
+      auto res = wrapped_tensor(t, res_indices, res_values)._coalesced_(true);
+      return std::make_tuple(res, static_cast<OptTensor>(indices_hash_sorted));
```
```suggestion
      return std::make_tuple(std::move(res), static_cast<OptTensor>(indices_hash_sorted));
```

nikitaved replied:
It is applied, I have no idea why github does not render that properly, however.

nikitaved replied:
It is applied, but I made a mistake and merged it into a PR higher in the stack...

Skylion007 commented on aten/src/ATen/native/sparse/SparseTensor.cpp:
```diff
@@ -744,6 +746,65 @@ SparseTensor _coalesce_sparse_cpu(const SparseTensor& self) {
+    }
+  }();
+
+  const auto rhs = mask.is_coalesced() ? wrapped_tensor(mask) : mask;
+
+  return std::make_tuple(lhs, rhs, lhs_hash_opt);
```
```suggestion
  auto rhs = mask.is_coalesced() ? wrapped_tensor(mask) : mask;

  return std::make_tuple(std::move(lhs), std::move(rhs), std::move(lhs_hash_opt));

nikitaved committed:
Update on "sparse.mm backward: performance improvements"

`torch.sparse.mm` - faster and without syncs in "most" cases.




cc alexsamardzic pearu cpuhrsch amjames bhosmer ezyang albanD zou3519 gqchen soulitzer Lezcano Varal7

[ghstack-poisoned]

nikitaved committed:
Update on "sparse.mm backward: performance improvements"

`torch.sparse.mm` - faster and without syncs in "most" cases.




cc alexsamardzic pearu cpuhrsch amjames bhosmer ezyang albanD zou3519 gqchen soulitzer Lezcano Varal7

[ghstack-poisoned]

nikitaved committed:
Update on "sparse.mm backward: performance improvements"

`torch.sparse.mm` - faster and without syncs in "most" cases.




cc alexsamardzic pearu cpuhrsch amjames bhosmer ezyang albanD zou3519 gqchen soulitzer Lezcano Varal7

[ghstack-poisoned]

nikitaved committed:
Update on "sparse.mm backward: performance improvements"

`torch.sparse.mm` - faster and without syncs in "most" cases.




cc alexsamardzic pearu cpuhrsch amjames bhosmer ezyang albanD zou3519 gqchen soulitzer Lezcano Varal7

[ghstack-poisoned]

nikitaved committed:
Update on "sparse.mm backward: performance improvements"

`torch.sparse.mm` - faster and without syncs in "most" cases.




cc alexsamardzic pearu cpuhrsch amjames bhosmer ezyang albanD zou3519 gqchen soulitzer Lezcano Varal7

[ghstack-poisoned]

nikitaved committed:
Update on "sparse.mm backward: performance improvements"

`torch.sparse.mm` - faster and without syncs in "most" cases.




cc alexsamardzic pearu cpuhrsch amjames bhosmer ezyang albanD zou3519 gqchen soulitzer Lezcano Varal7

[ghstack-poisoned]

nikitaved committed:
Update on "sparse.mm backward: performance improvements"

`torch.sparse.mm` - faster and without syncs in "most" cases.




cc alexsamardzic pearu cpuhrsch amjames bhosmer ezyang albanD zou3519 gqchen soulitzer Lezcano Varal7

[ghstack-poisoned]

nikitaved committed:
Update on "sparse.mm backward: performance improvements"

`torch.sparse.mm` - faster and without syncs in "most" cases.




cc alexsamardzic pearu cpuhrsch amjames bhosmer ezyang albanD zou3519 gqchen soulitzer Lezcano Varal7

[ghstack-poisoned]

cpuhrsch commented:
Do you have some timings on the performance improvements for this?

nikitaved commented:
@cpuhrsch , the main improvement of this PR is just removing syncs in the backward whenever possible. But here are some benchmarks:

Master:

```python
In [1]: import torch

In [2]: import itertools

In [3]: def make_input_grad(*shape):
   ...:     x = torch.rand(*shape, device='cuda')
   ...:     x_grad = torch.rand_like(x)
   ...:     return x.to_sparse(), x_grad.to_sparse()
   ...: 

In [4]: def run_test(*shape):
   ...:     for cflags in itertools.product(*itertools.repeat((True, False), 3)):
   ...:         x, grad = make_input_grad(*shape)
   ...:         y, _ = make_input_grad(*shape)
   ...:         x, y, grad = [t._coalesced_(f) for t, f in zip([x, y, grad], cflags)]
   ...:         x.requires_grad_(True)
   ...:         y.requires_grad_(True)
   ...:         print(f"x.coalesced: {cflags[0]}, y.coalesced: {cflags[1]}, grad.coalesced: {cflags[2]}")
   ...:         %timeit torch.autograd.grad(torch.sparse.mm(x, y), (x, y), (grad,)); torch.cuda.synchronize()
   ...: 

In [5]: run_test(32, 32)
x.coalesced: True, y.coalesced: True, grad.coalesced: True
1.32 ms ± 37.7 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
x.coalesced: True, y.coalesced: True, grad.coalesced: False
1.54 ms ± 524 ns per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
x.coalesced: True, y.coalesced: False, grad.coalesced: True
1.64 ms ± 7.69 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
x.coalesced: True, y.coalesced: False, grad.coalesced: False
1.9 ms ± 34.9 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
x.coalesced: False, y.coalesced: True, grad.coalesced: True
1.68 ms ± 37.5 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
x.coalesced: False, y.coalesced: True, grad.coalesced: False
1.88 ms ± 32.6 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
x.coalesced: False, y.coalesced: False, grad.coalesced: True
1.94 ms ± 16.1 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
x.coalesced: False, y.coalesced: False, grad.coalesced: False
2.16 ms ± 1.95 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

In [6]: run_test(64, 64)
x.coalesced: True, y.coalesced: True, grad.coalesced: True
1.52 ms ± 105 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
x.coalesced: True, y.coalesced: True, grad.coalesced: False
1.7 ms ± 5.5 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
x.coalesced: True, y.coalesced: False, grad.coalesced: True
1.79 ms ± 5.62 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
x.coalesced: True, y.coalesced: False, grad.coalesced: False
2.06 ms ± 23.5 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
x.coalesced: False, y.coalesced: True, grad.coalesced: True
1.78 ms ± 914 ns per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
x.coalesced: False, y.coalesced: True, grad.coalesced: False
2.02 ms ± 923 ns per loop (mean ± std. dev. of 7 runs, 100 loops each)
x.coalesced: False, y.coalesced: False, grad.coalesced: True
2.08 ms ± 1.11 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
x.coalesced: False, y.coalesced: False, grad.coalesced: False
2.32 ms ± 2.61 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

In [7]: run_test(128, 128)
x.coalesced: True, y.coalesced: True, grad.coalesced: True
2.18 ms ± 14.5 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
x.coalesced: True, y.coalesced: True, grad.coalesced: False
2.45 ms ± 1.03 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
x.coalesced: True, y.coalesced: False, grad.coalesced: True
2.53 ms ± 1.43 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
x.coalesced: True, y.coalesced: False, grad.coalesced: False
2.8 ms ± 901 ns per loop (mean ± std. dev. of 7 runs, 100 loops each)
x.coalesced: False, y.coalesced: True, grad.coalesced: True
2.53 ms ± 1.77 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
x.coalesced: False, y.coalesced: True, grad.coalesced: False
2.8 ms ± 1.15 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
x.coalesced: False, y.coalesced: False, grad.coalesced: True
2.93 ms ± 732 ns per loop (mean ± std. dev. of 7 runs, 100 loops each)
x.coalesced: False, y.coalesced: False, grad.coalesced: False
3.2 ms ± 1.12 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

This PR:
```python
In [5]: run_test(32, 32)
x.coalesced: True, y.coalesced: True, grad.coalesced: True
1.36 ms ± 41.3 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
x.coalesced: True, y.coalesced: True, grad.coalesced: False
1.5 ms ± 14.2 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
x.coalesced: True, y.coalesced: False, grad.coalesced: True
1.54 ms ± 499 ns per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
x.coalesced: True, y.coalesced: False, grad.coalesced: False
1.75 ms ± 939 ns per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
x.coalesced: False, y.coalesced: True, grad.coalesced: True
1.54 ms ± 669 ns per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
x.coalesced: False, y.coalesced: True, grad.coalesced: False
1.75 ms ± 921 ns per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
x.coalesced: False, y.coalesced: False, grad.coalesced: True
1.75 ms ± 1.84 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
x.coalesced: False, y.coalesced: False, grad.coalesced: False
1.96 ms ± 1.24 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)

In [6]: run_test(64, 64)
x.coalesced: True, y.coalesced: True, grad.coalesced: True
1.47 ms ± 89.5 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
x.coalesced: True, y.coalesced: True, grad.coalesced: False
1.67 ms ± 2.29 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
x.coalesced: True, y.coalesced: False, grad.coalesced: True
1.69 ms ± 1.97 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
x.coalesced: True, y.coalesced: False, grad.coalesced: False
1.92 ms ± 1.53 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
x.coalesced: False, y.coalesced: True, grad.coalesced: True
1.68 ms ± 909 ns per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
x.coalesced: False, y.coalesced: True, grad.coalesced: False
1.92 ms ± 7.13 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
x.coalesced: False, y.coalesced: False, grad.coalesced: True
1.94 ms ± 27.8 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
x.coalesced: False, y.coalesced: False, grad.coalesced: False
2.15 ms ± 722 ns per loop (mean ± std. dev. of 7 runs, 100 loops each)

In [7]: run_test(128, 128)
x.coalesced: True, y.coalesced: True, grad.coalesced: True
2.45 ms ± 54.3 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
x.coalesced: True, y.coalesced: True, grad.coalesced: False
2.41 ms ± 658 ns per loop (mean ± std. dev. of 7 runs, 100 loops each)
x.coalesced: True, y.coalesced: False, grad.coalesced: True
2.45 ms ± 2.44 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
x.coalesced: True, y.coalesced: False, grad.coalesced: False
2.72 ms ± 1.35 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
x.coalesced: False, y.coalesced: True, grad.coalesced: True
2.45 ms ± 2.3 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
x.coalesced: False, y.coalesced: True, grad.coalesced: False
2.72 ms ± 697 ns per loop (mean ± std. dev. of 7 runs, 100 loops each)
x.coalesced: False, y.coalesced: False, grad.coalesced: True
2.73 ms ± 879 ns per loop (mean ± std. dev. of 7 runs, 100 loops each)
x.coalesced: False, y.coalesced: False, grad.coalesced: False
3 ms ± 1.02 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

```

nikitaved committed:
Update on "sparse.mm backward: performance improvements"

`torch.sparse.mm` - faster and without syncs in "most" cases.




cc alexsamardzic pearu cpuhrsch amjames bhosmer ezyang albanD zou3519 gqchen soulitzer Lezcano Varal7

[ghstack-poisoned]

nikitaved commented:
@cpuhrsch , do you have any other comments? Otherwise it is ready to go.

nikitaved committed:
Update on "sparse.mm backward: performance improvements"

`torch.sparse.mm` - faster and without syncs in "most" cases.




cc alexsamardzic pearu cpuhrsch amjames bhosmer ezyang albanD zou3519 gqchen soulitzer Lezcano Varal7

[ghstack-poisoned]

nikitaved committed:
Update on "sparse.mm backward: performance improvements"

`torch.sparse.mm` - faster and without syncs in "most" cases.




cc alexsamardzic pearu cpuhrsch amjames bhosmer ezyang albanD zou3519 gqchen soulitzer Lezcano Varal7

[ghstack-poisoned]

nikitaved committed:
Update on "sparse.mm backward: performance improvements"

`torch.sparse.mm` - faster and without syncs in "most" cases.




cc alexsamardzic pearu cpuhrsch amjames bhosmer ezyang albanD zou3519 gqchen soulitzer Lezcano Varal7

[ghstack-poisoned]

nikitaved committed:
Update on "sparse.mm backward: performance improvements"

`torch.sparse.mm` - faster and without syncs in "most" cases.




cc alexsamardzic pearu cpuhrsch amjames bhosmer ezyang albanD zou3519 gqchen soulitzer Lezcano Varal7

[ghstack-poisoned]

nikitaved committed:
Update on "sparse.mm backward: performance improvements"

`torch.sparse.mm` - faster and without syncs in "most" cases.




cc alexsamardzic pearu cpuhrsch amjames bhosmer ezyang albanD zou3519 gqchen soulitzer Lezcano Varal7

[ghstack-poisoned]

nikitaved committed:
Update on "sparse.mm backward: performance improvements"

`torch.sparse.mm` - faster and without syncs in "most" cases.




cc alexsamardzic pearu cpuhrsch amjames bhosmer ezyang albanD zou3519 gqchen soulitzer Lezcano Varal7

[ghstack-poisoned]

nikitaved committed:
Update on "sparse.mm backward: performance improvements"

`torch.sparse.mm` - faster and without syncs in "most" cases.




cc alexsamardzic pearu cpuhrsch amjames bhosmer ezyang albanD zou3519 gqchen soulitzer Lezcano Varal7

[ghstack-poisoned]

cpuhrsch commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -6728,6 +6728,12 @@
     SparseCsrCPU, SparseCsrCUDA: sparse_mask_sparse_csr
   autogen: sparse_mask.out
 
+- func: _sparse_mask_projection(Tensor self, Tensor mask) -> Tensor
```
Are we sure this needs another native function or can this live in https://github.com/pytorch/pytorch/blob/master/torch/csrc/autograd/FunctionsManual.h/.cpp?

cpuhrsch replied:
@albanD - do you have some requirements here for this / can you take a look?

nikitaved replied:
The kernel uses dispatch stabs under the hood, hence the need for the native exterior.

nikitaved replied:
We would need more than a single kw argument for that, see https://github.com/pytorch/pytorch/pull/94991#discussion_r1172897761, i.e., the difference is in the direction of projection and/or whether intersecting matches require accumulation or not.

cpuhrsch commented:
@nikitaved - The timings look very promising!

It would be easier to compare as a table, but I looked through the timings and it seems like this version is consistently faster.

Great!

nikitaved committed:
Update on "sparse.mm backward: performance improvements"

`torch.sparse.mm` - faster and without syncs in "most" cases.




cc alexsamardzic pearu cpuhrsch amjames bhosmer ezyang albanD zou3519 gqchen soulitzer Lezcano Varal7

[ghstack-poisoned]

nikitaved committed:
Update on "sparse.mm backward: performance improvements"

`torch.sparse.mm` - faster and without syncs in "most" cases.




cc alexsamardzic pearu cpuhrsch amjames bhosmer ezyang albanD zou3519 gqchen soulitzer Lezcano Varal7

[ghstack-poisoned]

pearu commented on aten/src/ATen/native/sparse/SparseTensor.cpp:
```diff
@@ -816,6 +841,31 @@ SparseTensor sparse_mask(const Tensor& t, const SparseTensor& mask) {
+      mask.sizes());
+
+  if (!t.numel() || !t._nnz() || !mask.numel() || !mask._nnz()) {
+    auto res = t.clone();
```
Shouldn't `mask` be used instead of `t`? This is because the returned tensor should have the same indices as the `mask` has.

pearu replied:
So, do we allow the indices of `sparse_mask_like_grad` (and consequently `sparse_sparse_matmul_backward`) result to be variable: sometimes these are the indices of `x`, and sometimes the indices of `gx`?

nikitaved replied:
Yes, depending on what is projected onto what. To eliminated syncs, it is beneficial to project a sorted tensor onto a coalesced one.

pearu replied:
> No, it has different semantics.

What are the semantics differences between `_sparse_mask_projection` and `sparse_mask`?

nikitaved replied:
Projection is "reversed" and matches are not accumulated, as to not produce gradients which are multiples of the in-flowing gradients.

pearu replied:
@nikitaved this description is a little cryptic for me. Could explain the difference in terms of operation invariants rather than in the terms of a need (for which I have no doubts)?

For example, in the case of `sparse_mask(self, mask)`, we [have](https://pytorch.org/docs/main/generated/torch.Tensor.sparse_mask.html?highlight=sparse_mask#torch.Tensor.sparse_mask):
> Returns a new sparse tensor with values from a strided tensor `self` filtered by the indices of the sparse tensor mask. The values of `mask` sparse tensor are ignored. `self` and `mask` tensors must have the same shape.
> The returned sparse tensor might contain duplicate values if `mask` is not coalesced.
> The returned sparse tensor has the same indices as the sparse tensor `mask`.

How would you rephrase this documentation bit for `_sparse_mask_projection(mask)`?

nikitaved replied:
`self._sparse_mask_projection(mask)` will a tensor with the same indices as `self` and values which are nonzero for all matches between `self.indices()` and `mask.indices()`. The word "filter" is a bit confusing, as it does not specify the behavior of `sparse_mask`-like functions with uncoalesced inputs.

pearu replied:
Thanks, @nikitaved! So, IIUC, we have the following differences in-between `sparse_mask` and `_sparse_mask_projection`:
1. `sparse_mask` `input` and `mask` can have any layouts (`strided`, `sparse_coo`, `sparse_csr`, ... but `mask` must have some sparse layout), while `_sparse_mask_projection` supports only `sparse_coo` layout both in `input` and `mask`
2. `sparse_mask` result indices are the same as `mask` indices, while `_sparse_mask_projection` indices are the same as `input` indices

In other words, if `y = x.sparse_mask(mask)` and `z = x._sparse_mask_projection(mask)` then `z = x.sparse_mask(mask.sparse_mask(x))`. Is this correct?

nikitaved replied:
Agree with 1 and 2. The last point I disagree with (a mistake there?). These methods are never equivalent in general if at least one of the inputs is uncoalesced.

nikitaved replied:
If only `sparse_mask` did not require indices to be taken from `mask`, I would argue that `_sparse_mask_projection` is a better version of `sparse_mask` for sparse COO inputs, as it just does the filtering (match ? include : set to zero).

pearu replied:
OK, I see. `sparse_mask` has the weirdness of duplicating values when the mask is uncoalesced. So, something like the following should then hold:
`x._sparse_mask_projection(mask).coalesce()` is the same as `x.sparse_mask(mask.sparse_mask(x.coalesce()))`

nikitaved replied:
Yes, although `sparse_mask` sums up the duplicates, unlike `_sparse_mask_projection`. In this case we can for sure that as long as `mask` is coalesced, so is `res.sparse_mask(mask)` no matter whether `res` is coalesced or not.

nikitaved replied:
It is a doubling only if there are exactly two matches. A duplication would mean several memory copies and we do not have these.

pearu reviewed (changes requested):
Thanks, @nikitaved, for speed ups!
I have a number of nits and questions, but overall looks good.

nikitaved commented on aten/src/ATen/native/sparse/SparseTensor.cpp:
```diff
@@ -753,57 +814,21 @@ SparseTensor sparse_mask(const Tensor& t, const SparseTensor& mask) {
       mask.sizes());
 
-  if (!mask.numel()) {
+  if (!t.numel() || !mask.numel() || !mask._nnz()) {
```
```python
In [1]: import torch

In [2]: idx = torch.tensor([[0, 0, 0], [0, 1, 2]])

In [3]: val = torch.rand(3, 5, 0)

In [4]: rhs = torch.sparse_coo_tensor(idx, val, (5, 5, 5, 0))

In [5]: rhs
Out[5]: 
tensor(indices=tensor([[0, 0, 0],
                       [0, 1, 2]]),
       values=tensor([], size=(3, 5, 0)),
       size=(5, 5, 5, 0), nnz=3, layout=torch.sparse_coo)

```

nikitaved committed:
Update on "sparse.mm backward: performance improvements"

`torch.sparse.mm` - faster and without syncs in "most" cases.




cc alexsamardzic pearu cpuhrsch amjames bhosmer ezyang albanD zou3519 gqchen soulitzer Lezcano Varal7

[ghstack-poisoned]

nikitaved commented on aten/src/ATen/native/sparse/SparseTensor.cpp:
```diff
@@ -744,6 +746,65 @@ SparseTensor _coalesce_sparse_cpu(const SparseTensor& self) {
+}
 
 SparseTensor sparse_mask(const Tensor& t, const SparseTensor& mask) {
   TORCH_CHECK(
```
This check is needed because not all inputs are sparse, and `sparse_mask_like_prepare_sparse_inputs` works only with sparse inputs. I could remove the shape check in `sparse_mask_like_prepare_sparse_inputs` instead.

nikitaved commented on aten/src/ATen/native/sparse/SparseTensor.cpp:
```diff
@@ -816,6 +841,31 @@ SparseTensor sparse_mask(const Tensor& t, const SparseTensor& mask) {
+      " but mask has size ",
+      mask.sizes());
+
+  if (!t.numel() || !t._nnz() || !mask.numel() || !mask._nnz()) {
```
Fixed.

nikitaved committed:
Update on "sparse.mm backward: performance improvements"

`torch.sparse.mm` - faster and without syncs in "most" cases.




cc alexsamardzic pearu cpuhrsch amjames bhosmer ezyang albanD zou3519 gqchen soulitzer Lezcano Varal7

[ghstack-poisoned]

nikitaved committed:
Update on "sparse.mm backward: performance improvements"

`torch.sparse.mm` - faster and without syncs in "most" cases.




cc alexsamardzic pearu cpuhrsch amjames bhosmer ezyang albanD zou3519 gqchen soulitzer Lezcano Varal7

[ghstack-poisoned]

nikitaved commented on aten/src/ATen/native/sparse/SparseTensor.cpp:
```diff
@@ -816,6 +841,31 @@ SparseTensor sparse_mask(const Tensor& t, const SparseTensor& mask) {
+  TORCH_INTERNAL_ASSERT(mask.is_sparse());
+
+  TORCH_CHECK(
+      mask.sizes().equals(t.sizes()),
```
Eliminated inside `sparse_mask_like_prepare_sparse_inputs`.

nikitaved committed:
Update on "sparse.mm backward: performance improvements"

`torch.sparse.mm` - faster and without syncs in "most" cases.




cc alexsamardzic pearu cpuhrsch amjames bhosmer ezyang albanD zou3519 gqchen soulitzer Lezcano Varal7

[ghstack-poisoned]

nikitaved committed:
Update on "sparse.mm backward: performance improvements"

`torch.sparse.mm` - faster and without syncs in "most" cases.




cc alexsamardzic pearu cpuhrsch amjames bhosmer ezyang albanD zou3519 gqchen soulitzer Lezcano Varal7

[ghstack-poisoned]

nikitaved committed:
Update on "sparse.mm backward: performance improvements"

`torch.sparse.mm` - faster and without syncs in "most" cases.




cc alexsamardzic pearu cpuhrsch amjames bhosmer ezyang albanD zou3519 gqchen soulitzer Lezcano Varal7

[ghstack-poisoned]

nikitaved committed:
Update on "sparse.mm backward: performance improvements"

`torch.sparse.mm` - faster and without syncs in "most" cases.




cc alexsamardzic pearu cpuhrsch amjames bhosmer ezyang albanD zou3519 gqchen soulitzer Lezcano Varal7

[ghstack-poisoned]

nikitaved committed:
Update on "sparse.mm backward: performance improvements"

`torch.sparse.mm` - faster and without syncs in "most" cases.




cc alexsamardzic pearu cpuhrsch amjames bhosmer ezyang albanD zou3519 gqchen soulitzer Lezcano Varal7

[ghstack-poisoned]

nikitaved committed:
Update on "sparse.mm backward: performance improvements"

`torch.sparse.mm` - faster and without syncs in "most" cases.




cc alexsamardzic pearu cpuhrsch amjames bhosmer ezyang albanD zou3519 gqchen soulitzer Lezcano Varal7

[ghstack-poisoned]

pearu commented on aten/src/ATen/native/sparse/SparseTensor.cpp:
Thanks for adding this comment!

This function returns three tensors. How do these relate to the inputs and to the result of the projection?

pearu replied:
IIUC, for
```python
lhs, rhs, lhs_hash_opt = sparse_mask_like_prepare_sparse_inputs(t, mask)
```
we have the following invariants
```python
lhs.indices() == t.indices()[lhs_hash]
lhs.values() == t.values()[lhs_hash]
rhs == mask
```
where `lhs_hash = lhs_hash_opt or slice(None)`, `lhs` may be a copy or a view of `t`, `rhs` is uncoaleced view of `mask`.

nikitaved replied:
See the comment in https://github.com/pytorch/pytorch/pull/94991#discussion_r1177575459.

pearu commented on aten/src/ATen/native/sparse/SparseTensor.cpp:
Is using `t._nnz()` safe considering that `res_values.size(0) == t._nnz()` is not always true?

pearu commented on aten/src/ATen/native/sparse/SparseTensor.cpp:
What is the advantage of returning `mask` as uncoalesced?

nikitaved replied:
If `mask` is uncoalesced, but the other argument is not, the COO intersection kernel will return a tensor with the same indices as `mask` but will do binary search of `mask` hashes into the hashes of the other arguments all without calls to `sort` and `coalesce`. The COO intersection kernel is heavily optimized to take advantage of `is_coalesced` of either arguments.

pearu replied:
This (forcing a coalesced input to be uncoalesced) sounds like a convoluted way to control COO intersection kernel functionality (to ensure that the result indices are the input indices).

nikitaved replied:
It still better than writing things from scratch, imho. The COO intersection kernel is not a public function with fixed semantics.

pearu reviewed (commented):
I have a few clarification questions, otherwise LGTM!

nikitaved committed:
Update on "sparse.mm backward: performance improvements"

`torch.sparse.mm` - faster and without syncs in "most" cases.




cc alexsamardzic pearu cpuhrsch amjames bhosmer ezyang albanD zou3519 gqchen soulitzer Lezcano Varal7

[ghstack-poisoned]

nikitaved committed:
Update on "sparse.mm backward: performance improvements"

`torch.sparse.mm` - faster and without syncs in "most" cases.




cc alexsamardzic pearu cpuhrsch amjames bhosmer ezyang albanD zou3519 gqchen soulitzer Lezcano Varal7

[ghstack-poisoned]

nikitaved committed:
Update on "sparse.mm backward: performance improvements"

`torch.sparse.mm` - faster and without syncs in "most" cases.




cc alexsamardzic pearu cpuhrsch amjames bhosmer ezyang albanD zou3519 gqchen soulitzer Lezcano Varal7

[ghstack-poisoned]

nikitaved commented:
@pearu , do you have any other concerns? I will provide comfortable interfaces to COO intersection kernels as a follow-up to indeed reduce cognitive burden...

pearu reviewed (approved):
LGTM! Thanks, @nikitaved!

cpuhrsch commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -1440,19 +1464,13 @@ Tensor sparse_sparse_matmul_backward(
   if (grad_order == 0) {
     auto a_grad = _sparse_sparse_matmul(grad, b.conj().t());
-    return a_grad.mul(mask_ones_like(a.coalesce()));
+    return sparse_mask_like_grad(a, a_grad);
```
Ok, so essentially we're starting to write manual fusions. If we had torch.compile support we could presumably generate the code for this pattern. Since this is a hot path this is ok, but it's going to start to become difficult to test correctness.

Since `_sparse_mask_projection` is a native function you could even write a test that compares it to the pattern that you're fusing. 

Is this a common general pattern? Maybe we can apply it in more locations. If we generalize the fused composite native function that you're adding here a bit, maybe it applies to more locations? It can then also subsume all the logic of `sparse_mask_like_grad` and we can explicitly say "This function represents the fusion of this pattern". That'll be easier to understand to future maintainers.

nikitaved replied:
@cpuhrsch , alternatively, we can make this function composite implicit now that `sparse_mask` does support backward with COO inputs. This way we can remove the code for backward altogether. The result: much less code, still sync-less backward albeit potentially slower compared to this impl (there is only one way to call backward, there is no back-and-force between grad and input, because of the fixed semantics of `sparse_mask`). But, yes, it is a generic pattern forces by some "sparse-semantics" functions which might become obsolete when differentiable `sparse_mask`.
Another possibility: we can create a public function which is doing either `_sparse_mask_projection` and/or `sparse_mask` (whichever is faster) which is explicitly differentiable and could be used to enforce sparse semantics all while having flexible backward.

cpuhrsch replied:
I think we should go with your first approach for now and keep the amount of code low and then work on torch.compile support to provide these fusions. We can keep what you have for now as a reference and goal for that integration. Otherwise we'll have to undo these manual fusions again once we have torch.compile support.

cpuhrsch reviewed (changes requested):
I think this looks good, I'm just worried that future maintainers might find this difficult to grok quickly. What if we reframe this work as implementing a fused function for a specific pattern of operations?

nikitaved commented:
@cpuhrsch , do you mind merging this for now? Once `sparse_mask` backward is in, I will make sparse-mm composite implicit and we will likely remove a LOT of already in the codebase code...

nikitaved committed:
Update on "sparse.mm backward: performance improvements"

`torch.sparse.mm` - faster and without syncs in "most" cases.




cc alexsamardzic pearu cpuhrsch amjames bhosmer ezyang albanD zou3519 gqchen soulitzer Lezcano Varal7

[ghstack-poisoned]

cpuhrsch commented:
@pytorchbot merge
